File: cmd/feed-reader/url_parser.go
package main

import (
	"strings"

	"github.com/nakatanakatana/feed-reader/store"
)

// URLParser extracts user info from URLs based on rules.
type URLParser struct {
	rules []store.UrlParsingRule
}

func NewURLParser(rules []store.UrlParsingRule) *URLParser {
	return &URLParser{rules: rules}
}

// ExtractUserInfo attempts to extract user information from the given URL.
func (p *URLParser) ExtractUserInfo(urlStr string) *store.ExtractedUserInfo {
	// Simple implementation: check each rule.
	// For better performance with many rules, we could index by domain.
	for _, rule := range p.rules {
		if rule.RuleType == "subdomain" {
			// Expected pattern: <user>.<rule.Pattern>
			// Example: rule.Pattern = "example.com", url = "https://user1.example.com/post"
			// Extract domain part from URL first
			domainPart := getDomainFromURL(urlStr)
			if strings.HasSuffix(domainPart, "."+rule.Pattern) {
				user := strings.TrimSuffix(domainPart, "."+rule.Pattern)
				if user != "" && !strings.Contains(user, ".") {
					return &store.ExtractedUserInfo{
						User:   user,
						Domain: rule.Pattern,
					}
				}
			}
		} else if rule.RuleType == "path" {
			// Expected pattern: <rule.Pattern>/<user>
			// Example: rule.Pattern = "domain.com/users", url = "https://domain.com/users/user1/post"
			if strings.Contains(urlStr, "://"+rule.Pattern+"/") {
				parts := strings.Split(urlStr, "://"+rule.Pattern+"/")
				if len(parts) > 1 {
					userPart := parts[1]
					user := strings.Split(userPart, "/")[0]
					if user != "" {
						// Extract domain from pattern
						domain := strings.Split(rule.Pattern, "/")[0]
						return &store.ExtractedUserInfo{
							User:   user,
							Domain: domain,
						}
					}
				}
			}
		}
	}
	return nil
}

func getDomainFromURL(urlStr string) string {
	parts := strings.Split(urlStr, "://")
	if len(parts) < 2 {
		return ""
	}
	remaining := parts[1]
	domain := strings.Split(remaining, "/")[0]
	// Remove port if present
	domain = strings.Split(domain, ":")[0]
	return domain
}

File: frontend/src/lib/url-parser.ts
import type { URLParsingRule } from "../gen/item/v1/item_pb";

export interface ExtractedUserInfo {
  user: string;
  domain: string;
}

export class URLParser {
  constructor(private rules: URLParsingRule[]) {}

  extractUserInfo(urlStr: string): ExtractedUserInfo | null {
    try {
      const url = new URL(urlStr);
      const domainPart = url.hostname;

      for (const rule of this.rules) {
        if (rule.ruleType === "subdomain") {
          // Expected pattern: <user>.<rule.pattern>
          // Example: rule.pattern = "example.com", url = "https://user1.example.com/post"
          if (domainPart.endsWith(`.${rule.pattern}`)) {
            const user = domainPart.slice(0, -(rule.pattern.length + 1));
            if (user !== "" && !user.includes(".")) {
              return {
                user,
                domain: rule.pattern,
              };
            }
          }
        } else if (rule.ruleType === "path") {
          // Expected pattern: <rule.pattern>/<user>
          // Example: rule.pattern = "domain.com/users", url = "https://domain.com/users/user1/post"
          const fullPath = url.hostname + url.pathname;
          if (fullPath.startsWith(`${rule.pattern}/`)) {
            const remaining = fullPath.slice(rule.pattern.length + 1);
            const user = remaining.split("/")[0];
            if (user !== "") {
              // Extract domain from pattern
              const domain = rule.pattern.split("/")[0];
              return {
                user,
                domain,
              };
            }
          }
        }
      }
    } catch (_e) {
      return null;
    }
    return null;
  }
}
